### ISSUE TITLE
Dask runtime for Modin takes longer than expected

### ISSUE THREAD

**Issue by devin-petersohn** [modin-project/modin#289]
We have a Dask runtime for Modin thanks to @coobas and @pcahyna 🎉🎉🎉 in #281.

Running some benchmarks, it seems to take longer than I expected, but I could not find any problems with the implementation. This will take some deeper profiling.

We have a python-only backend that does not offer any parallelism and the `dask.delayed` implementation is running slower than that. This is likely not a Dask performance problem, rather the architecture should be updated in places to support the Dask runtime better.

We may also want to consider using Dask Futures.

cc @mrocklin (I thought you would be interested to see that Dask is now supported 🎊)

**Comment by mrocklin** [comment]
Glad to hear it!  Nice works @coobas and @pcahyna 

Looking very briefly at #281 it looks like it's common to use dask.delayed and then dask.compute one after another in quick succession.  If this is the case then it might not be optimal.  Typically with dask.delayed you want to build larger graphs and then compute them all at once.  The current approach (if I understand it correctly) does a few computations and then blocks until they finish, then does a few more and blocks, etc..  If my understanding is correct then the futures interface might be a better fit.

If you have a benchmark that I can run I'd be happy to run something.  I recommend running the dask.distributed scheduler so that you can see the dashboard during execution.  This is pretty useful for diagnostics.  http://docs.dask.org/en/latest/diagnostics-distributed.html

**Comment by devin-petersohn** [comment]
Benchmarking bash script: 
https://github.com/modin-project/modin/blob/master/.jenkins/performance-tests/run_perf_test.sh

The evaluation script is run with `pytest`:

https://github.com/modin-project/modin/blob/master/modin/pandas/test/performance-tests/test_performance.py

We have Jenkins set up to automatically test the performance of each PR, and here is the console output of the most recent build:

https://amplab.cs.berkeley.edu/jenkins/job/Modin-Performance-PRB/93/console

Let me know how it goes!

**Comment by coobas** [comment]
We are glad we've made this contribution., thanks to @devin-petersohn for the very smooth on-boarding.

We noticed the performance was not great ... With our knowledge of modin we could not figure out though how to use dask in a better way. What we basically assumed was that modin needs to apply its functions to `pandas.DataFrame`'s when requested. Ideally, as @mrocklin pointed out, we would have just a lazy call graph constructed by modin which would be `compute`d only once the final result is needed. Right now `dask.delayed` is used more as a (distributed) storage plus a shallow lazy evaluation.

**Comment by coobas** [comment]
Btw, we looked into [Futures](http://docs.dask.org/en/latest/futures.html#futures) but since
> This interface is good for arbitrary task scheduling like dask.delayed, but is immediate rather than lazy, 

we chose `delayed` because we though that the lazy evaluation would fit better. Again, this decision was based on a very limited knowledge of modin.

**Comment by devin-petersohn** [comment]
Modin was originally designed for Ray, which has no lazy library. Internally in Modin, we have structures in place to (soon) allow for lazy operations and query planning in non-lazy environments. I am glad that we have a benchmark for `dask.delayed`, as it gives a good baseline for performance once we have a query planner in place.

I think we can use both. For the `dask.delayed` component, we probably want to use `dask.compute(...)` instead of `obj.compute()` on each partition. This is something that will need to changing in the internals of Modin. There are probably some other things we can do, but this is the lowest hanging fruit to improve `dask.delayed` performance.

**Comment by mrocklin** [comment]
Just to be clear, using dask.delayed and dask.compute in every operation is pretty suboptimal.  I wouldn't expect good performance, and I definitely wouldn't expect any scalability.

**Comment by devin-petersohn** [comment]
Thanks @mrocklin! Just so there's not any misunderstanding, here is when we are using `get` (which is what invokes `dask.compute`):

* Any operation that results in a `Series` (only because it is not distributed yet)
* Any operation that calls `__repr__` or `__str__`, and then only on the partitions getting printed
* An operation that defaults to pandas implementation.
* An operation that converts the DataFrame into another structure (e.g. `as_matrix`)
* Writing to a file

`compute` should only be called in those conditions, and based on the implementation by @coobas et. al. I believe that is what is happening, which is good. I also tested running something that does not trigger a `compute` internally (e.g. `df.abs()`) and it returns immediately as expected. I do not believe the issue is with the implementation, rather that Modin is still missing enough of the pandas API to make the performance difference noticeable.

There are still some missing pieces in the `modin.pandas` API that may result in `compute` being called frequently in certain workloads. This means that the issue is that we need to keep building out the API of Modin to better use the parallelism in `dask.delayed`, though I would also like to add Dask Futures to Modin as well.

**Comment by coobas** [comment]
Hi, @pcahyna pushed a fix to what we believe was a bug in our contribution to modin. @devin-petersohn , plese have a look at the last four commits (a1a1f6c^) in https://github.com/coobas/modin/commits/AxisPartition . All tests were passing for  use only with this fix.

**Comment by devin-petersohn** [comment]
Thanks @coobas!

It looks great, feel free to PR. There will likely be a merge conflict because of #290. The structure of the code has slightly changed, but I was able to group the project by the execution engine (`modin.engines.dask`) instead of by the abstract class. This is a cleaner way of keeping the files together.

I will gladly help you merge #290 with the new file structure. I also saw some interesting PRs on the @coobas fork related to additions that could be made to the Dask backend. They look like they would be exciting additions!
